proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1203 @@
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        [i32, i32, i32]
+      else:
+        [i32]
+
+# Determines the core wasm types used to represent a variant's cases. This
+# overlays various cases into the same list of types, so the length of types is
+# the length of the longest case, and types are "unified" to a common
+# representation to ensure there's enough bits to hold enough values for all
+# cases.
+def flatten_cases(direction, cases):
+  result = []
+  for case in cases:
+    if case:
+      tys = flatten(direction, case)
+      for i, ty in enumerate(tys):
+        if i < len(result):
+          result[i] = unify(result[i], ty)
+        else:
+          result.push(ty)
+  return result
+
+def unify(a, b):
+  match (a, b)
+    (i64, _) |
+    (_, i64) |
+    (i32, f64) |
+    (f64, i32) => i64,
+
+    (i32, i32) |
+    (i32, f32) |
+    (f31, i32) => i32,
+
+    (f32, f32) => f32,
+    (f64, f64) | (f32, f64) | (f64, i32) => f64,
+
+def num_packed_i32s(fields):
+  return align_to(fields.len(), 32) / 32
+```
+
+The only difference in exports/imports currently is that `push-buffer` and
+`pull-buffer` are represented with a singular `i32` instead of three `i32`
+values. This is because if an export is receiving a buffer it's coming from
+somewhere else which means it's a handle to a buffer and various "intrinsics",
+listed below, will be used to operate on the handle.
+
+## Lifting
+
+The `lift` function here takes four parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lifting is happening: lifting an export means lifting
+  a result, lifting an import means lifting a parameter.
+* `src` - this is an instance where the value is being lifted from. This could
+  be a WebAssembly instance or the host itself.
+* `ty` - the interface type that we're lifting into
+* `values` - the core wasm values iterator we have from the function signature
+  or results that we're lifting from. Note that due to type-checking and the
+  canonical ABI for function signatures it's guaranteed that `values` has the
+  right types to lift into `ty`.
+
+This defines how an interface types value is decoded from linear memory and
+WebAssembly values. Note that this function is modeled as a generator. This is
+used later during the definition of `fuse` to model how lifting/lowering is
+interleaved.
+
+```
+def lift(direction, src, ty, values):
+  match ty
+    f32 => yield(values.assert_next(wasm_f32)),
+    f64 => yield(values.assert_next(wasm_f64)),
+
+    s8 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= -128 && val <= 127)
+      yield(val as s8)
+    }
+
+    s16 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= -32768 && val <= 32767)
+      yield(val as s16)
+    }
+
+    u8 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= 0 && val <= 255)
```

Comment:
This is missing a way of specifying that the incoming `i32` value is interpreted as unsigned. The `assert_next` call is the same as in the `s8` case. It's clear in the s8/u8 cases what's intended, but it's less clear in some of the cases below like `char` and `list`.

Reply:
Ah good point, I've tried to add in some `as XX` to be explicit about how the values are interpreted